src/client/datascience/jupyter/kernels/kernelConnectionWrapper.ts

```diff
@@ -44,84 +44,145 @@ export class KernelConnectionWrapper implements Kernel.IKernelConnection {
     public readonly unhandledMessage = new Signal<this, IMessage<MessageType>>(this);
     public readonly anyMessage = new Signal<this, Kernel.IAnyMessageArgs>(this);
     public get serverSettings() {
-        return this.kernelConnection.serverSettings;
+        return (this.kernelConnection || this._previousKernelConnection).serverSettings;
     }
     public readonly disposed = new Signal<this, void>(this);
+    private _kernelConnection!: Kernel.IKernelConnection;
+    private _previousKernelConnection!: Kernel.IKernelConnection;
+    // private _isRestarting?: boolean;
+    private get kernelConnection(): undefined | Kernel.IKernelConnection {
```

Comment:
When restarting a kernel, the `kernel.IKernelConnection` gets changed, hence we cannot store this in a variable, as it changes.

Also, when a kernel is shutdown, this could be null.

This makes the code work as expected (when used by 3rd party extensions) when restarting kernels or shutting them down.

Reply:
Not sure why on restart we don't just update ```_kernelConnection``` during a restart?

Reply:
Thats exactly what we do today.
This is required, because while restart is taking place the underlying object `kernel.session.kernel` could have already changed. Hence even before restart has completed we could end up with a whole new kernel connection, so any code running here will end up accessing the old item and things could fall over (basically thats what i found while debugging).

Hence the check to see if `kernel.session.kernel !== _kernelConnection`.

Reply:
Sorry so the use case is:
- User runs a kernel
- 3rd party uses kernel
- User restarts
- During restart 3rd party accesses the IKernelConnection and it's undefined?
- Crash?

Can we not just have a restart event on the kernel? It seems like the 3rd party is using the kernel when it shouldn't be? It's the same scenario if they try using a kernel before one is started.

Reply:
> During restart 3rd party accesses the IKernelConnection and it's undefined?

At this point it could have already been updated even before the whole restart has completed (i.e. our extension restart has not yet completed, the async code is still running).

> It's the same scenario if they try using a kernel before one is started.

Basically we're hooked into events and the like and all of those event handlers (for getting data from the kernel would be out dated).
3rd party doesn't even know that a kernel is restarting all they're doing is accesing a kernel, they're not necessarily restarting the kernel (the user could be restarting the kernel using the jupyter extension)

> It seems like the 3rd party is using the kernel when it shouldn't be?

Not true, they don't know its restarting.
E.g. i could write code to get kernel events or widget messages, but if in the middle of all of that the kernel is restarted, then tha'ts when things fall over (as the 3rd party extension code is accessing old kernle when its no longer valid).
The problem is we're still exposing the old kernel when its not valid.

Reply:
> They don't know it's restarting

This seems like the root of the problem. We're attempting to let the other extension keep a hold of this wrapper for the IKernelConnection. I think they need to watch for status change events and stop holding onto it when it dies. 

Otherwise the state of this thing won't make sense. 

And if they need to know when a restart is happening, then we need to provide an event for that.

Reply:
Thats wrong, in the kernel API today they can hold onto the kernel connection. its totally valid.
In the jupyter lab API you can hold onto the object and restart.

The whole point of the wrapper is to provide the same behaviour: 
* the kernel wrapper is a singleton (looks like the same kernel connection object)
* When restaring they still deal with the same object (that's how jupyter works)

So the problem isn't the 3rd party extenxion , but us, we need to ensure we provide the same thing & they don't need to worry about any restarts or the like, tha'ts not how the API works today, we don't do this either, hence 3rd party extensions dont need to do this

Reply:
So if the goal is to behave like jupyter, then we need to modify the status change events then. Otherwise a restart will look like a death of a kernel.

Reply:
>  Otherwise a restart will look like a death of a kernel.

No it won't, we don't kill the old kernel before swapping.
I did want to send restart events as well, but its just too messy to fix that, will get that sorted later in another debt week.

I've tested this and it works as expected. Only problem is we don't get `restarting` events. Which isn't too bad for now